tests/src/main/java/net/neoforged/neoforge/debug/transfer/VanillaHandlersTests.java

```diff
@@ -0,0 +1,566 @@
+
+            if (resourceHandler.insert(2, book, 1, tx) != 1) throw helper.assertionException("Should have inserted 1 book");
+            if (bookshelf.getLastInteractedSlot() != 2) throw helper.assertionException("Last modified slot should be 2");
+
+            if (resourceHandler.extract(1, book, 1, tx) != 1) throw helper.assertionException("Should have extracted 1 book");
+            if (bookshelf.getLastInteractedSlot() != 1) throw helper.assertionException("Last modified slot should be 1");
+
+            // Now, create an aborted nested transaction.
+            try (Transaction nested = Transaction.open(tx)) {
+                if (resourceHandler.insert(book, 100, nested) != 5) throw helper.assertionException("Should have inserted 5 books");
+                // Now, last modified slot should be 5.
+                if (bookshelf.getLastInteractedSlot() != 5) throw helper.assertionException("Last modified slot should be 5");
+            }
+
+            // And it's back to 1 in theory.
+            if (bookshelf.getLastInteractedSlot() != 1) throw helper.assertionException("Last modified slot should be 1");
+            tx.commit();
+        }
+
+        if (bookshelf.getLastInteractedSlot() != 1) throw helper.assertionException("Last modified slot should be 1 after committing transaction");
+
+        // Let's also check the state properties. Only slot 2 should be occupied.
+        BlockState state = bookshelf.getBlockState();
+
+        if (state.getValue(BlockStateProperties.CHISELED_BOOKSHELF_SLOT_0_OCCUPIED)) throw helper.assertionException("Slot 0 should not be occupied");
+        if (state.getValue(BlockStateProperties.CHISELED_BOOKSHELF_SLOT_1_OCCUPIED)) throw helper.assertionException("Slot 1 should not be occupied");
+        if (!state.getValue(BlockStateProperties.CHISELED_BOOKSHELF_SLOT_2_OCCUPIED)) throw helper.assertionException("Slot 2 should be occupied");
+        if (state.getValue(BlockStateProperties.CHISELED_BOOKSHELF_SLOT_3_OCCUPIED)) throw helper.assertionException("Slot 3 should not be occupied");
+        if (state.getValue(BlockStateProperties.CHISELED_BOOKSHELF_SLOT_4_OCCUPIED)) throw helper.assertionException("Slot 4 should not be occupied");
+        if (state.getValue(BlockStateProperties.CHISELED_BOOKSHELF_SLOT_5_OCCUPIED)) throw helper.assertionException("Slot 5 should not be occupied");
+
+        helper.succeed();
+    }
+
+    @GameTest
+    @EmptyTemplate
+    @TestHolder(description = "Tests that shulker boxes cannot be inserted into other shulker boxes.")
+    public static void testShulkerNoInsert(ExtendedGameTestHelper helper) {
+        BlockPos pos = new BlockPos(0, 2, 0);
+        helper.setBlock(pos, Blocks.SHULKER_BOX);
+        ShulkerBoxBlockEntity shulker = helper.getBlockEntity(pos, ShulkerBoxBlockEntity.class);
+        var resourceHandler = new WorldlyContainerWrapper(shulker, null);
```

Comment:
Do we want to be testing it isn't just failing to be inserted as a side effect of it using the null side?

Reply:
Changed it to test all sides.